source/common/common/thread.h

```diff
@@ -172,23 +172,34 @@ class AtomicPtr : private AtomicPtrArray<T, 1, alloc_mode> {
 struct MainThread {
   using MainThreadSingleton = InjectableSingleton<MainThread>;
   bool inMainThread() const { return main_thread_id_ == std::this_thread::get_id(); }
-  static void init() { MainThreadSingleton::initialize(new MainThread()); }
-  static void clear() {
-    delete MainThreadSingleton::getExisting();
-    MainThreadSingleton::clear();
-  }
-  static bool isMainThread() {
-    // If threading is off, only main thread is running.
-    if (MainThreadSingleton::getExisting() == nullptr) {
-      return true;
-    }
-    // When threading is on, compare thread id with main thread id.
-    return MainThreadSingleton::get().inMainThread();
+  bool inTestThread() const {
+    return test_thread_id_.has_value() && (test_thread_id_.value() == std::this_thread::get_id());
   }
+  void registerTestThread() { test_thread_id_ = std::this_thread::get_id(); }
+  void registerMainThread() { main_thread_id_ = std::this_thread::get_id(); }
+  static bool initialized() { return MainThreadSingleton::getExisting() != nullptr; }
+  static void initMainThread();
+  static void initTestThread();
+  static void clear();
```

Comment:
Can we have some doxygen comments for all these methods?

In particular, I think it's only safe to call clear() when threads are quiescent, e.g. between test methods. I'm not sure where it's actually called. Otherwise I see a few races in isMainThread().

Reply:
We call initMainThread() and clear() inside the constructor and destructor of ThreadLocal::InstanceImpl, which is guaranteed to be in main thread. (before threading is initialized and after threading has been shut down.

Reply:
From a system perspective I believe you.

From an internal API perspective, I am just asking you to document the assumptions in method comments, so that in the future, for reasons we can't predict, no one adds a thread-unsafe call to clear() during operation.

Reply:
Thanks, comments added.